### PR TITLE
Merge in IAP JWT verification nginx module.

### PR DESCRIPTION
**Pull request by kmbarry1**
Implementation of an nginx module to do automatic validation of IAP JWT assertions.

Not necessarily the final form, but has all the required functionality and is reasonably well-tested.

Not yet integrated with ESP, so checkin is a null op.

### CODE REVIEW

**Comment by drewinglis** [comment]
Is this a copy/paste of existing code, or is this a new implementation? (I.e. has this code been reviewed before?) If the former, could you link the existing implementation? Thanks.

**Comment by kmbarry1** [comment]
This is entirely new implementation. Due to my novice GitHub abilities, it ended up being a single commit here. No prior review.

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    return false;
+  }
+
+  std::unique_ptr<Jwt> jwt =
+      Jwt::parse_from_string(std::string(raw_jwt, raw_jwt_len));
+  if (jwt == nullptr) {
+    return false;
+  }
+
+  if (now < jwt->iat() - 60 || now > jwt->exp() + 60) {
```
Comment (drewinglis) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: 60 -> CLOCK_SKEW, I assume.

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+         header_dot_payload.length(),
+         digest);
+  size_t der_sig_len;
+  std::unique_ptr<const uint8_t []> der_sig = jose_sig_to_der_sig(
+      jwt.signature_bytes(), jwt.signature_length(), &der_sig_len);
+  if (der_sig == nullptr) {
+    return false;
+  }
+
+  if (ECDSA_verify(0,
```
Comment (drewinglis) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: I see from the docs that this first parameter is ignored. It would be good if you could add a comment to that effect.

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    return false;
+  }
+
+  return true;
+}
+
+// Most of this implementation is copied from the BoringSSL code that creates
+// DER-encoded signatures in the first place--see crypto/dsa/dsa_asn1.c in the
+// BoringSSL codebase.
+std::unique_ptr<uint8_t []> jose_sig_to_der_sig(
```
Comment (drewinglis) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: `uint8_t[]`

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  if (CBB_add_asn1(&der_sig_cbb, &child, CBS_ASN1_SEQUENCE) != 1
+      || BN_marshal_asn1(&child, r.get()) != 1
+      || BN_marshal_asn1(&child, s.get()) != 1
+      || CBB_flush(&der_sig_cbb) != 1) {
+    CBB_cleanup(&der_sig_cbb);
+    return nullptr;
+  }
+
+  *der_sig_len = CBB_len(&der_sig_cbb);
+  std::unique_ptr<uint8_t []> der_sig(new (std::nothrow) uint8_t[*der_sig_len]);
```
Comment (drewinglis) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: `uint8_t[]`

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification_test.cc [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+// iat: 1001
+// exp: 1600
+// aud: /projects/1234/apps/some-app-id
+// email: nobody@example.com
+// iss: https://cloud.google.com/iap
+// sub: accounts.google.com:12345
+constexpr char VALID_JWT[] = "eyJhbGciOiJFUzI1NiIsImtpZCI6IjFhIiwidHlwIjoiSldUIn0.eyJhdWQiOiIvcHJvamVjdHMvMTIzNC9hcHBzL3NvbWUtYXBwLWlkIiwiZW1haWwiOiJub2JvZHlAZXhhbXBsZS5jb20iLCJleHAiOjE2MDAsImlhdCI6MTAwMSwiaXNzIjoiaHR0cHM6Ly9jbG91ZC5nb29nbGUuY29tL2lhcCIsInN1YiI6ImFjY291bnRzLmdvb2dsZS5jb206MTIzNDUifQ.Oez6MwXTCLuJ-U8VHVgiHtADGv818ZMQjnjOQSgUJ9NU1UcLT5ZyT8DqgbAnRLuv0Tf9hdmcBilouJD46oPo2w";
+constexpr size_t VALID_JWT_LEN = strlen(VALID_JWT);
+
+// A time at which VALId_JWT should be valid.
```
Comment (drewinglis) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification_test.cc]: VALId -> VALID

File: third_party/iap_jwt_verify_nginx/src/jwt_test.cc [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+namespace iap {
+namespace {
+
+using ::google::cloud::iap::test::url_safe_b64_encode;
+
+TEST(JwtTest, ValidJwtParsesSuccessfully) {
+  const std::string encoded_header =
+      "eyJhbGciOiJFUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6InhkS193dyJ9";
+  const std::string encoded_payload =
+      "eyJpc3MiOiJodHRwczovL2Nsb3VkLmdvb2dsZS5jb20vaWFwIiwic3ViIjoiYWNjb3VudHMuZ29vZ2xlLmNvbToxMjM0NSIsImVtYWlsIjoibm9ib2R5QGV4YW1wbGUuY29tIiwiYXVkIjoiL2EvYi9jL2QiLCJleHAiOjYwMCwiaWF0IjowfQ";
```
Comment (drewinglis) [inline third_party/iap_jwt_verify_nginx/src/jwt_test.cc]: To respect line length limits, split into multiline string like so:

    const std::string encoded_payload =
        "asdfasdfasdfadsf...asdf"
        "asdfsdafsadfsdf...asdf";

File: third_party/iap_jwt_verify_nginx/src/jwt_test.cc [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+// SUCH DAMAGE.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "src/jwt.h"
+
+#include "gtest/gtest.h"
+
```
Comment (drewinglis) [inline third_party/iap_jwt_verify_nginx/src/jwt_test.cc]: Remove newline.

File: third_party/iap_jwt_verify_nginx/src/module.cc [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+      if (part->next == nullptr) {
+        break;
+      }
+
+      part = part->next;
+      h = reinterpret_cast<ngx_table_elt_t *>(part->elts);
+      i = 0;
+    }
+
+    static const ngx_str_t header_name = ngx_string(IAP_JWT_HEADER_NAME);
```
Comment (drewinglis) [inline third_party/iap_jwt_verify_nginx/src/module.cc]: Move this outside the for loop.

File: third_party/iap_jwt_verify_nginx/src/module.cc [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+
+  time_t now = ngx_time();
+  if (now >= main_conf->last_iap_state_check.load()
+                + main_conf->iap_state_cache_time_sec) {
+    if (ngx_trylock(&iap_state_lock, 1)) {
+      // Check the condition again to account for the case where another thread
+      // acquired and released the lock between when we checked the condition
+      // and when we acquired the lock.
+      if (now >=
+            main_conf->last_iap_state_check.load()
```
Comment (drewinglis) [inline third_party/iap_jwt_verify_nginx/src/module.cc]: Move this up to the previous line.

File: third_party/iap_jwt_verify_nginx/test/BUILD [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    ],
+    deps = [
+        ":iap_jwt_generation",
+        "@jsoncpp//:lib_json",
+    ],
+    data = [
+        ":keys",
+    ],
+    copts = [
+        "-Iexternal/jsoncpp/jsoncpp-1.8.0/include",
```
Comment (drewinglis) [inline third_party/iap_jwt_verify_nginx/test/BUILD]: Refactor into a constant.

File: third_party/iap_jwt_verify_nginx/test/BUILD [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+    srcs = [
+        "key_generator.cc",
+    ],
+    deps = [
+        ":iap_jwt_generation",
+        "@boringssl//:crypto",
+        "@jsoncpp//:lib_json",
+    ],
+    copts = [
+        "-Iexternal/boringssl/include",
```
Comment (drewinglis) [inline third_party/iap_jwt_verify_nginx/test/BUILD]: Refactor into a constant.

File: third_party/iap_jwt_verify_nginx/README.md [hunk 12]
```diff
@@ -0,0 +1,2 @@
+An nginx module to verify IAP JWTs (signed assertions that identify the user
+and scope to the intended receiving application).
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/README.md]: Done.

File: third_party/iap_jwt_verify_nginx/gtest.BUILD [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+#
+################################################################################
+#
+# Used to build the GoogleTest library.
+
+cc_library(
+    name = "main",
+    srcs = glob(
+        ["googletest-release-1.8.0/googletest/src/*.cc"],
+	exclude = ["googletest-release-1.8.0/googletest/src/gtest-all.cc"]
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/gtest.BUILD]: Done.

File: third_party/iap_jwt_verify_nginx/gtest.BUILD [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    ),
+    hdrs = glob([
+        "googletest-release-1.8.0/googletest/include/**/*.h",
+	"googletest-release-1.8.0/googletest/src/*.h"
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/gtest.BUILD]: Done.

File: third_party/iap_jwt_verify_nginx/jsoncpp.BUILD [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+# Used to build the JsonCpp library.
+
+cc_library(
+    name = "lib_json",
+    srcs = glob(
+        ["jsoncpp-1.8.0/src/lib_json/*.cpp"],
+    ),
+    hdrs = glob([
+        "jsoncpp-1.8.0/include/json/*.h",
+	"jsoncpp-1.8.0/src/lib_json/*.h",
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/jsoncpp.BUILD]: Done.

File: third_party/iap_jwt_verify_nginx/jsoncpp.BUILD [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+	"jsoncpp-1.8.0/src/lib_json/*.inl"
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/jsoncpp.BUILD]: Done.

File: third_party/iap_jwt_verify_nginx/src/BUILD [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+#
+################################################################################
+
+gtest_include_copt = (
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/BUILD]: Done.

File: third_party/iap_jwt_verify_nginx/src/BUILD [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+    "-Iexternal/gtest/googletest-release-1.8.0/googletest/include")
+
+cc_library(
+    name = "ngx_iap_jwt_verify",
+    srcs = [
+        "module.h",
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/BUILD]: This target must have public visibility (it's what consumers of this module must link against). See nginx_modules.c under test/ (and the nginx-iap cc_binary there) for an example of how this is done.

Consumers don't need to include module.h, however, so I put it under srcs instead of hdrs. ESP does it the same way:
https://github.com/cloudendpoints/esp/blob/master/src/nginx/BUILD#L89

File: third_party/iap_jwt_verify_nginx/src/BUILD [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+        "NGX_IAP_JWT_VERIFY",
+    ],
+    deps = [
+        ":iap_jwt_verification",
+        "@boringssl//:crypto",
+        "@jsoncpp//:lib_json",
+        "@nginx//:http",
+    ],
+    copts = [
+        "-Iexternal/boringssl/include",
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/BUILD]: Done.

File: third_party/iap_jwt_verify_nginx/src/BUILD [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+        "-Iexternal/jsoncpp/jsoncpp-1.8.0/include",
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/BUILD]: Done.

File: third_party/iap_jwt_verify_nginx/src/BUILD [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+        "-Iexternal/jsoncpp/jsoncpp-1.8.0/src/lib_json",
+    ],
+)
+
+cc_library(
+    name = "iap_jwt_verification",
+    srcs = [
+        "base64_util.cc",
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/BUILD]: Why? I don't see what that accomplishes besides cluttering the BUILD file.

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc (hunk 1 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: Ah thank you, I made the constant then forgot to use it...

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc (hunk 2 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: Done.

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc (hunk 3 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: Done.

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc (hunk 4 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: Done.

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification_test.cc (hunk 5 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification_test.cc]: Done.

File: third_party/iap_jwt_verify_nginx/src/jwt_test.cc (hunk 7 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/jwt_test.cc]: Done.

File: third_party/iap_jwt_verify_nginx/src/jwt_test.cc (hunk 6 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/jwt_test.cc]: Done.

File: third_party/iap_jwt_verify_nginx/src/module.cc (hunk 8 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/module.cc]: Done. Also capitalized it to reflect its const status.

File: third_party/iap_jwt_verify_nginx/src/module.cc (hunk 9 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/module.cc]: Done.

File: third_party/iap_jwt_verify_nginx/test/BUILD (hunk 10 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/test/BUILD]: Done.

File: third_party/iap_jwt_verify_nginx/test/BUILD (hunk 11 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/test/BUILD]: Done.

**Comment by kmbarry1** [comment]
Thanks for the comments! All addressed but one that I wanted to ask a quick question about.

For dealing with the size of the review, I was going to suggest that I guide you through a few files at a time, as the factorization is clean enough to allow this. Start with base64_util{.h, .cc, _test.cc}, when we're happy with those we'll move on to the next chunk.

A few other things I should mention:
1) I've tried to stick exclusively to C++11 features and earlier for various reasons. If you want to discuss more, let me know.

2) The files module.h and module.cc follow aesthetic and idiomatic conventions that differ from the rest of the files, for consistency with how nginx code looks/functions, since I have to implement a lot of nginx interfaces there...all other files are devoid of nginx code references, and have a more modern C++ feel.

3) If you aren't familiar with nginx module writing, we should probably sync up so I can give you an outline of how things work before you review module{.h, .cc} and the integration tests (these will be the last things to review though, so it doesn't need to be right away).

File: third_party/iap_jwt_verify_nginx/src/BUILD (hunk 18 above)
Comment (drewinglis) [inline third_party/iap_jwt_verify_nginx/src/BUILD]: Sounds good.

File: third_party/iap_jwt_verify_nginx/src/base64_util.cc [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+namespace iap {
+
+bool max_decoded_length(const std::string &b64_str, size_t *len) {
+  size_t b64_str_len = b64_str.length();
+  *len = 3 * (b64_str_len / 4);
+  switch (b64_str_len % 4) {
+    case 0:
+      break;
+    case 1:
+      return false;
```
Comment (Cynocracy) [inline third_party/iap_jwt_verify_nginx/src/base64_util.cc]: is_valid or something like that? (please doc)

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.h [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+#define NGINX_IAP_JWT_VERIFY_IAP_JWT_VERIFICATION_H
+
+#include "src/jwt.h"
+#include "src/types.h"
+
+namespace google {
+namespace cloud {
+namespace iap {
+
+bool iap_jwt_is_valid(const char *raw_jwt,
```
Comment (Cynocracy) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.h]: Document some of the params here

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.h [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+                      const char *expected_aud,
+                      size_t expected_aud_len,
+                      std::shared_ptr<iap_key_map_t> keys);
+
+bool verify_iap_jwt_sig(const Jwt &jwt, const iap_key_map_t &keys);
+
+std::unique_ptr<uint8_t[]> jose_sig_to_der_sig(
+    const uint8_t *jose_sig, size_t jose_sig_len, size_t *der_sig_len);
+
+std::shared_ptr<iap_key_map_t> load_keys(const char *file_name,
```
Comment (Cynocracy) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.h]: same, document

File: third_party/iap_jwt_verify_nginx/test/jwt_generation.h [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+// SUCH DAMAGE.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef NGINX_IAP_JWT_VERIFY_TEST_JWT_GENERATION_H
+#define NGINX_IAP_JWT_VERIFY_TEST_JWT_GENERATION_H
+
+#include <memory>
+
+#include "json/json.h"
```
Comment (Cynocracy) [inline third_party/iap_jwt_verify_nginx/test/jwt_generation.h]: Is this file still needed?

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.h (hunk 23 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.h]: Good catch, I forgot to document these *facepalm*.

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.h (hunk 24 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.h]: Done.

File: third_party/iap_jwt_verify_nginx/test/jwt_generation.h (hunk 25 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/test/jwt_generation.h]: I'm not sure whether you're referring to the "json/json.h" include or this file as a whole, but...yes, both are needed. 

Assuming you meant the latter: though I don't actively generate JWTs in any of the tests,
1) it would be unfortunate to have to rewrite JWT generation code if we ever needed more
2) I'd like to keep open the option of building this test infrastructure out to support tests that leverage dynamic JWT generation (e.g. stress testing, or just fancier integration tests)

**Comment by kmbarry1** [comment]
Comments addressed, thanks! Hold off on merging until I see if someone from my team wants to review.

File: third_party/iap_jwt_verify_nginx/test/BUILD [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+
+JSONCPP_COPTS_FLAGS = [
+    "-Iexternal/boringssl/include",
+    "-Iexternal/jsoncpp/jsoncpp-1.8.0/include",
+]
+
+BSSL_INCLUDE_COPT = "-Iexternal/boringssl/include"
+
+py_test(
+    name = "nginx_test",
```
Comment (matthewg) [inline third_party/iap_jwt_verify_nginx/test/BUILD]: Do you need to do anything special s.t. this actually gets run as a presubmit or what-not?

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc [hunk 27]
```diff
@@ -0,0 +1,201 @@
...
+#include "openssl/sha.h"
+#include "src/jwt.h"
+#include "src/key_util.h"
+
+namespace google {
+namespace cloud {
+namespace iap {
+
+// Let's refuse to even process JWTs larger than a certain size.
+constexpr size_t MAX_RAW_JWT_SIZE = 4096;
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: Yep, I checked the typical length first...this is safely bigger than any JWT we're likely to issue even if we increase our key size or add more fields to the JWT, but it saves us the stomach ulcers of wondering "What happens if someone malicious sends us a JWT the size of the moon?".

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc [hunk 28]
```diff
@@ -0,0 +1,201 @@
...
+
+// Allow for one minute of clock skew when verifying IAP JWT iat/exp values.
+constexpr uint64_t CLOCK_SKEW = 60;
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: Hmmm, it might even make sense to not validate iat at all. I can't think of a case where validating iat is actually important for security.

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc [hunk 29]
```diff
@@ -0,0 +1,201 @@
...
+  }
+
+  if (!verify_iap_jwt_sig(*jwt, *keys)) {
+    return false;
+  }
+
+  return true;
+}
+
+bool verify_iap_jwt_sig(const Jwt &jwt, const iap_key_map_t &keys) {
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: Right, the algorithm is baked-in pretty hard. I'll add a comment calling this out, but I think it would be a waste of cycles to check the 'alg' value as things stand today.

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc [hunk 30]
```diff
@@ -0,0 +1,201 @@
...
+                   key_iter->second.get()) != 1) {
+    return false;
+  }
+
+  return true;
+}
+
+// Most of this implementation is copied from the BoringSSL code that creates
+// DER-encoded signatures in the first place--see crypto/dsa/dsa_asn1.c in the
+// BoringSSL codebase.
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: Frankly I find the CBB API confusing and cryptic, but I figured it was still better than building ASN.1 structures by hand.

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc [hunk 31]
```diff
@@ -0,0 +1,201 @@
...
+std::unique_ptr<uint8_t[]> jose_sig_to_der_sig(
+    const uint8_t *const jose_sig,
+    const size_t jose_sig_len,
+    size_t *const der_sig_len) {
+  if (jose_sig_len != 2*FINITE_FIELD_BYTE_SIZE) {
+    return nullptr;
+  }
+
+  // Convert the two components of the jose signature into BIGNUMs
+  bssl::UniquePtr<BIGNUM> r(
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: Yeah, I think this signature interconversion issue is a big part of why there weren't any good JWT libraries I could use in C/C++ that support ES256.

File: third_party/iap_jwt_verify_nginx/src/jwt.cc [hunk 32]
```diff
@@ -0,0 +1,209 @@
+// Copyright (C) 2002-2016 Igor Sysoev
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/jwt.cc]: I looked into alternatives, but almost no C/C++ JWT libraries support ES256, the one I did find was neither actively maintained (last commit: 2 years ago) nor widely used, and there are trust issues with pulling in random code for a security feature. Not to mention you have to be careful about licensing. A JWT processing nginx module does exist, but it's 1) not open source or free (requires a commercial subscription), and 2) doesn't seem to have the functionality we want (no evidence that it supports ES256, doesn't have built-in audience validation so we'd still have to do that ourselves).

While rolling my own was the last thing I wanted to do, I felt it was the best choice after I investigated all the reuse options.

File: third_party/iap_jwt_verify_nginx/src/module.cc [hunk 33]
```diff
@@ -0,0 +1,541 @@
...
+
+// IAP JWT verification handler--core business logic lives here and in functions
+// called by this handler.
+ngx_int_t ngx_http_iap_jwt_verification_handler(ngx_http_request_t *r) {
+  ngx_iap_jwt_verify_loc_conf_t *loc_conf =
+      reinterpret_cast<ngx_iap_jwt_verify_loc_conf_t *>(
+          ngx_http_get_module_loc_conf(r, ngx_iap_jwt_verify_module));
+
+  if (loc_conf->iap_jwt_verify == 0) {
+    return NGX_OK;
```
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/module.cc]: Nope, iap_jwt_verify is the flag that indicates whether IAP JWT enforcement should be done for a given location. If it equals zero, enforcement is not being done, which means access is allowed. This handler's job, if called, is to make an access decision--and if no enforcement is being done, then the decision has to be "OK, access allowed".

I'll add a comment to the code.

File: third_party/iap_jwt_verify_nginx/test/BUILD (hunk 26 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/test/BUILD]: I'll admit, I don't even know how one would set that up in Github (presumably using Travis CI somehow). I don't think we need to block this merge on figuring that out, but it's definitely something to solve before we deploy this.

**Comment by kmbarry1** [comment]
Thanks Matthew, comments replied to/addressed, please take a look.

File: third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc (hunk 30 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc]: Update: found a better way to do this.

File: third_party/iap_jwt_verify_nginx/src/jwt.cc (hunk 32 above)
Comment (kmbarry1) [inline third_party/iap_jwt_verify_nginx/src/jwt.cc]: Update: apparently I was Googling slightly-wrong search terms all along because it seems there are a couple decent options I didn't find until just now:

https://github.com/latchset/jose
https://github.com/cisco/cjose

Neither one is used internally at Google AFAICT, so they still don't meet my bar for "I'm comfortable pulling these into a security feature without extensive review". We could certainly considering migrating to one (most likely the Cisco one, it's endorsed by OpenID).
